test/gtest/p4runtime.cpp

```diff
@@ -519,6 +522,7 @@ TEST_F(P4Runtime, P4_16_MatchFields) {
         { 36, "lShift", 16, MatchField::TERNARY },
         { 37, "plusSix", 16, MatchField::EXACT },
         { 38, "plusSix", 16, MatchField::TERNARY },
+        { 39, "h.hUnion.c.yetAnotherHeaderField", 8, MatchField::EXACT },
```

Comment:
That is hardly a sufficient test. You should check that the `type_name` field is set properly and that there is a corresponding entry in the `new_types` map in P4Info.

Reply:
This will take sometime to write this code since I do not have documentation on this code.  I can add a TODO and complete the code in a future PR.  I have addressed rest of the comments - please take a look - thanks.

Reply:
Additionally, there are numerous .p4 tests added in this PR where `type` has been used and new_types exist for the `type` in p4info.

Reply:
> That is hardly a sufficient test. You should check that the `type_name` field is set properly and that there is a corresponding entry in the `new_types` map in P4Info.

Please see the latest commit to this PR.  How did you intend to set `type_name` in the  P4_16_MatchFields test P4 code?  All one can do is define the `type` and use the type as the table MatchField.  Thus, in the `expected` MatchFields it seems contrived to set null strings and one string for type_name.